src/bridge/cockpitpcpmetrics.c

```diff
@@ -0,0 +1,1017 @@
+            return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static void
+send_object (CockpitPcpMetrics *self,
+             JsonObject *object)
+{
+  CockpitChannel *channel = (CockpitChannel *)self;
+  GBytes *bytes;
+
+  bytes = cockpit_json_write_bytes (object);
+  cockpit_channel_send (channel, bytes, TRUE);
+  g_bytes_unref (bytes);
+}
+
+static void
+send_array (CockpitPcpMetrics *self,
+            JsonArray *array)
+{
+  CockpitChannel *channel = (CockpitChannel *)self;
+  GBytes *bytes;
+  JsonNode *node;
+  gsize length;
+  gchar *ret;
+
+  node = json_node_new (JSON_NODE_ARRAY);
+  json_node_set_array (node, array);
+  ret = cockpit_json_write (node, &length);
+  json_node_free (node);
+
+  bytes = g_bytes_new_take (ret, length);
+  cockpit_channel_send (channel, bytes, TRUE);
+  g_bytes_unref (bytes);
+}
+
+static JsonObject *
+build_meta (CockpitPcpMetrics *self,
+            pmResult *result)
+{
+  JsonArray *instances;
+  JsonArray *units;
+  JsonArray *types;
+  JsonArray *semantics;
+  JsonArray *array;
+  JsonObject *root;
+  pmValueSet *vs;
+  gint64 timestamp;
+  char *instance;
+  int i, j;
+  int rc;
+
+  timestamp = (result->timestamp.tv_sec * 1000) +
+              (result->timestamp.tv_usec / 1000);
+
+  root = json_object_new ();
+  json_object_set_int_member (root, "timestamp", timestamp);
+  json_object_set_int_member (root, "interval", self->interval);
+
+  instances = json_array_new ();
+  for (i = 0; i < result->numpmid; i++)
+    {
+      vs = result->vset[i];
+
+      /* When negative numval is an error code ... we don't care */
+      if (vs->numval < 0 || (vs->numval == 1 && vs->vlist[0].inst == PM_IN_NULL))
+        {
+          json_array_add_null_element (instances);
+        }
+      else
+        {
+          array = json_array_new ();
+
+          for (j = 0; j < vs->numval; j++)
+            {
+              /* PCP guarantees that the result is in the same order as requested */
+              rc = pmNameInDom (self->metrics[i].desc.indom, vs->vlist[j].inst, &instance);
+              if (rc != 0)
+                {
+                  g_warning ("%s: instance name lookup failed: %s", self->name, pmErrStr (rc));
+                  instance = NULL;
+                }
+
+              /* HACK: We can't use json_builder_add_string_value here since
+                 it turns empty strings into 'null' values inside arrays.
+
+                 https://bugzilla.gnome.org/show_bug.cgi?id=730803
+              */
+              {
+                JsonNode *string_element = json_node_alloc ();
+                json_node_init_string (string_element, instance);
+                json_array_add_element (array, string_element);
+              }
+
+              if (instance)
+                free (instance);
+            }
+          json_array_add_array_element (instances, array);
+        }
+    }
+  json_object_set_array_member (root, "instances", instances);
+
+  units = json_array_new ();
```

Comment:
Again, I think the form of this meta data should match the form that it is specified in the open command "metrics" option.